### PR TITLE
Split out IndexedField trait

### PR DESCRIPTION
**Pull request by randomPoison**
* Move the `count`, `index`, `add_list`, and `get_list` methods into a separate `IndexedField` trait.
* Remove unnecessary `count` and `index` methods from collection types :tada:
* Use more efficient `Schema_Get*List` methods for primitive types.

---

I'm making this a draft PR because I want to write some tests around the usage of `Schem_Add*List`. Those methods don't copy the list, they just store a pointer back to it, so we need to make sure we enforce the lifetime constraints correctly. I suspect the current implementation isn't doing that correctly, and I'll want to make any changes to fix that as part of this PR.

### CODE REVIEW

**Comment by randomPoison** [comment]
Okay, so I think I've confirmed that the current implementation of `add_list` is unsound. The underlying C API is storing a pointer to the buffer owned by the source `Vec`, but we're not doing anything to prevent that source data from being dropped while the schema object still exists. I've added a compile fail test that shows that unsound code still compiles.

I'll start looking into what changes we can make to fix this unsoundness. I suspect the solution is to add a lifetime parameter to all schema data types indicating the lifetime of any data they borrow, e.g. `SchemaObject<'a>`. My main concern with that approach is that it may have a negative effect on ergonomics and may conflict with otherwise-valid code.

I'm also curious about the exact details around how the lifetimes work when passing the schema data to the SDK. For example, say that I'm building up a `Schema_ComponentData` to send as part of a worker op, and in populating the underlying `Schema_Object` I use `Schema_AddInt32List` to populate one of the fields. If I call `Worker_Connection_SendCreateEntityRequest` (for example) and pass it the `Schema_ComponentData` (as part of the `Worker_ComponentData`), at what point can I be sure that the SDK is done with the borrowed array of data? Is that `Schema_ComponentData` going to be stored in some way and used at a later point, or will it be used synchronously in the function call such that I'm okay to free that data once the function returns? @dgavedissian I think you'd know the answer to that one best.

**Comment by jamiebrynes7** [comment]
On using the `Schema_Add*List` methods. For reference, the C# interop SDK does the following to get around the lifetime issues: 

```csharp
        /// <summary>
        /// Functions that append a list of primitive values for a particular field ID to an object. Note
        /// that, for best performance, fields should be added to the object in field ID order.
        /// This function will allocate a copy of the supplied <c>values</c> parameter.
        /// </summary>
        public void AddFloatList(FieldId fieldId, float[] values)
        {
            Uint32 bufferLength = (Uint32) (values.Length * Marshal.SizeOf(typeof(float)));
            Uint8* buffer = AllocateBuffer(bufferLength);
            fixed (float* valueBuffer = values)
            {
                CWorker.Memcpy(buffer, (Uint8*) valueBuffer, (UIntPtr) bufferLength);
            }

            AddFloatList(fieldId, (float*) buffer, values.Length);
        }
```

Where `AllocateBuffer` is:

```csharp
        /// <summary>
        /// Allocates a buffer of the specified length in bytes from memory owned by this SchemaObject
        /// instance. The memory is freed by a call to the Destroy() function associated with the parent object.
        /// </summary>
        /// <remark>
        /// Note: this is useful for allocating memory that must live as long as the root schema type
        /// instance, for example to pass to <see cref="MergeFromBuffer(byte*, Uint32)"/>.
        /// </remark>
        public Uint8* AllocateBuffer(Uint32 bufferLength) => CSchema.AllocateBuffer(handle, bufferLength);
```

TL;DR - Allocate a buffer of the required size which has its lifetime tied to the `SchemaObject` (by the Worker SDK) and memcpy the data into that buffer, then use that buffer as the "list".

**Comment by randomPoison** [comment]
Ah, that's it! I knew @dgavedissian had previously explained to me how to solve this problem, but I couldn't remember what the solution was. It should be a simple fix too, so I'll get that updated :grin:

**Comment by randomPoison** [comment]
Okay, that should be fixed now! The logic for allocating the buffer does manual work for aligning the data, since I don't think `Schema_AllocateBuffer` guarantees any kind of alignment (at least the docs don't say anything about alignment), so I'd appreciate extra attention there to make sure I got it right 😄

File: spatialos-sdk/src/worker/schema/primitives.rs [hunk 1]
```diff
@@ -56,29 +77,54 @@ macro_rules! impl_primitive_field {
...
-            }
+            fn add_list(object: &mut SchemaObject, field: FieldId, data: &[$rust_type]) {
+                // Determine how large the buffer needs to be.
+                //
+                // NOTE: We allocate extra padding when allocating the buffer in order to have room
+                // to adjust the alignment of the buffer to match the alignment of `Self::RustType`
+                // and still have enough room in the buffer for the right number of elements.
+                let byte_len = {
+                    let data_len = data.len() * mem::size_of::<Self::RustType>();
+                    let padding = mem::align_of::<Self::RustType>() - 1;
```
[the remark this answers is not in the document]
Comment (randomPoison) [reply spatialos-sdk/src/worker/schema/primitives.rs]: The idea is that if we can't assume that the buffer returned by `Schema_AllocateBuffer` is properly aligned for our type, we may need to "shift" the buffer by up to `alignment - 1` bytes in order to get proper alignment. We therefore need to allocate an extra `alignment - 1` bytes of padding in order to ensure we have enough space in the buffer to copy all of the data after we've fixed the alignment.

... as I'm writing this, it occurs to me that this is *probably* not the right thing to do. For example, if we leave 3 bytes of padding at the beginning of the buffer in order to get proper alignment, there's probably going to be 3 bytes of garbage at the beginning of the buffer when the data is received by another worker. Any worker trying to read this data would have to somehow know how much padding was added. We could try to encode that in an extra byte at the beginning of the buffer, but that really feels like something I'd expect to either be documented as a necessary step or be handled by the SDK directly. @dgavedissian do you know if there are any alignment guarantees around `Schema_AllocateBuffer`? I feel like I've asked this before, and I wouldn't be surprised if it guaranteed large enough alignment to be valid for all primitive types.

[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply spatialos-sdk/src/worker/schema/primitives.rs]: >  For example, if we leave 3 bytes of padding at the beginning of the buffer in order to get proper alignment, there's probably going to be 3 bytes of garbage at the beginning of the buffer when the data is received by another worker.

I'm not sure this is necessarily true, since we pass in the pointer to the aligned buffer to the `Schema_AddxxxList` method. So the garbage bytes shouldn't ever be read by the schema API, since the buffer isn't directly sent over the wire, it's just a allocation that has its lifetime tied to the schema object.

[the remark this answers is not in the document]
Comment (randomPoison) [reply spatialos-sdk/src/worker/schema/primitives.rs]: Huh, that's a good point. So maybe its fine after all? Is there a good example of how this is handled in the SDKs for other languages? That would be a good way to confirm how to correctly handle this, I think.

[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply spatialos-sdk/src/worker/schema/primitives.rs]: Pasting the two implementations side-by-side

**C#**
```csharp
        /// <summary>
        /// Functions that append a list of primitive values for a particular field ID to an object. Note
        /// that, for best performance, fields should be added to the object in field ID order.
        /// This function will allocate a copy of the supplied <c>values</c> parameter.
        /// </summary>
        public void AddFloatList(FieldId fieldId, float[] values)
        {
            Uint32 bufferLength = (Uint32) (values.Length * Marshal.SizeOf(typeof(float)));
            Uint8* buffer = AllocateBuffer(bufferLength);
            fixed (float* valueBuffer = values)
            {
                CWorker.Memcpy(buffer, (Uint8*) valueBuffer, (UIntPtr) bufferLength);
            }

            AddFloatList(fieldId, (float*) buffer, values.Length);
        }
```

**Rust**

```rust
            fn add_list(object: &mut SchemaObject, field: FieldId, data: &[$rust_type]) {
                // Determine how large the buffer needs to be.
                //
                // NOTE: We allocate extra padding when allocating the buffer in order to have room
                // to adjust the alignment of the buffer to match the alignment of `Self::RustType`
                // and still have enough room in the buffer for the right number of elements.
                let byte_len = {
                    let data_len = data.len() * mem::size_of::<Self::RustType>();
                    let padding = mem::align_of::<Self::RustType>() - 1;
                    data_len + padding
                };

                // Allocate a buffer that is owned by `object`.
                let buffer = unsafe {
                    let data_ptr = Schema_AllocateBuffer(object.as_ptr_mut(), byte_len as _);
                    let buffer = slice::from_raw_parts_mut(data_ptr, byte_len);

                    // Convert the byte buffer into a correctly-aligned slice of the data type.
                    let (_prefix, buffer, _suffix) = buffer.align_to_mut::<Self::RustType>();
                    buffer
                };

                // Populate the buffer.
                buffer.copy_from_slice(data);

                unsafe {
                    $schema_add_list(
                        object.as_ptr_mut(),
                        field,
                        buffer.as_ptr() as *const _,
                        buffer.len() as u32,
                    );
                }
            }
```
Two things I noticed: 

* There is no alignment in the C# version. That difference may be that we directly copy the memory across (`CWorker.Memcpy` aliases the C `memcpy` function). 
* `from_raw_parts_mut` requires `data_ptr` to be aligned in the contract and we do the alignment _after_ we call this. 

---

Had a quick peek at the implementation and the memory is allocated with `std::max_align_t` as the alignment (https://en.cppreference.com/w/cpp/types/max_align_t) which means we shouldn't need to worry about the alignment if I'm reading the cpp docs correctly: 

> std::max_align_t is a trivial standard-layout type whose alignment requirement is at least as strict (as large) as that of every scalar type.

[the remark this answers is not in the document]
Comment (randomPoison) [reply spatialos-sdk/src/worker/schema/primitives.rs]: Ah! I swear one of you two previously told me that it uses `max_align_t` under the hood 🤦‍♀. Thinking about it, though, we could probably do a similar copy in the Rust version. Currently we convert the raw buffer into a mutable slice of the appropriate type, and then copy over to that slice. Instead, we could convert the input data into a byte slice and copy *that* into the allocated buffer. I think doing it that way would be both simpler and would keep the Rust version more consistent with the other SDKs.

[the remark this answers is not in the document]
Comment (randomPoison) [reply spatialos-sdk/src/worker/schema/primitives.rs]: Eh, okay I say that *but* it's actually much easier to convert the byte buffer to a typed slice than to go the other way around 🤔 So I'm actually going just remove the alignment logic and leave a comment noting that it's safe because the C SDK guarantees max alignment.
